pkg/controller/direct/compute/firewallpolicyrule/firewallpolicyrule_controller.go

```diff
@@ -0,0 +1,310 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package firewallpolicyrule
+
+import (
+	"context"
+	"fmt"
+
+	gcp "cloud.google.com/go/compute/apiv1"
+	computepb "cloud.google.com/go/compute/apiv1/computepb"
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/compute/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const ctrlName = "firewallpolicyrule-controller"
+
+func init() {
+	registry.RegisterModel(krm.ComputeFirewallPolicyRuleGVK, NewFirewallPolicyRuleModel)
+}
+
+func NewFirewallPolicyRuleModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &firewallPolicyRuleModel{config: config}, nil
+}
+
+type firewallPolicyRuleModel struct {
+	config *config.ControllerConfig
+}
+
+// model implements the Model interface.
+var _ directbase.Model = &firewallPolicyRuleModel{}
+
+type firewallPolicyRuleAdapter struct {
+	firewallPolicy         string
+	priority               int64
+	firewallPoliciesClient *gcp.FirewallPoliciesClient
+	desired                *krm.ComputeFirewallPolicyRule
+	actual                 *computepb.FirewallPolicyRule
+	reader                 client.Reader
+}
+
+var _ directbase.Adapter = &firewallPolicyRuleAdapter{}
+
+func (m *firewallPolicyRuleModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
```

Comment:
I do not have externalRef for this resource yet, this resource does not have unique identifier, and since it's a field of FirewallPolicy, I do not think there'll be other resource reference to it. No sure if we want to add create an externalRef for virtual resource like this.

Reply:
Maybe we can use the firewall policy external ref + "/rules/" + priority